### ISSUE TITLE
SupportVectorMachine learn error

### ISSUE THREAD

**Issue by kiran-reddy1** [accord-net/framework#368]
Could you please tell me the solution for this issue? I am using sequential optimization.

```
            var learn = new SequentialMinimalOptimization<Gaussian>()
            {
                UseComplexityHeuristic = true,
                UseKernelEstimation = true
            };

            SupportVectorMachine<Gaussian> svm = learn.Learn(iMatrix, tMatrix);
```

You can see the input iMatrix first few values in the captured picture in the locals in visual studio.
The tMatrix (output) is one dimensional array with values 1 and -1.

![capture-svm](https://cloud.githubusercontent.com/assets/14985336/21751868/583d01a0-d59c-11e6-839b-aa0694dd3a04.PNG)

**Comment by cesarsouza** [comment]
Hi there,

Normally the SVM or learning algorithm should have already checked for that, but can you please confirm that your iMatrix does not contain any element that is '''null'''?

Regards,
Cesar

**Comment by kiran-reddy1** [comment]
Hi Cesar,
I have truncated my array to less than the original size. Now i am not getting the error. But when my target is less than zero i got errors here:  
          double error = new ZeroOneLoss(outPuts).Loss(predicted);
           double loss = new CategoryCrossEntropyLoss(outPuts).Loss(prob);
So i have changed my target to 0 instead of -1. Now it works fine.  Also i am facing problems to use the Accord.Net framework in visual studio 2013 and 2015. It's throwing errors.

**Comment by cesarsouza** [comment]
Hi there,

If your classification problem is binary, you may want to use the [BinaryCrossEntropyLoss](http://accord-framework.net/docs/html/T_Accord_Math_Optimization_Losses_BinaryCrossEntropyLoss.htm) instead of the CategoryCrossEntropy one. And I had also recently fixed a problem with the ZeroOneLoss in the latest release (3.4.0, released yesterday).

Regarding the problems with Visual Studio, that's very weird. I actually develop the framework on VS2013, VS2015 and VS2017 RC, and I don't have problems. Are you using any VS extensions, such as for example, ReSharper?

If yes, you might want to try [opening Visual Studio with extensions disabled by passing the /SafeMode command line argument to devenv.exe](https://msdn.microsoft.com/en-us/library/ms241278(v=vs.140).aspx) (i.e. ```devenv.exe /SafeMode```).

Regards,
Cesar

**Comment by kiran-reddy1** [comment]
Thanks Cesar,

I updated my Nuget package. Now the previously working code throwing error
which is highlighted in the yellow colour:

 var teacher = new MulticlassSupportVectorLearning()
            {
                // Configure the learning algorithm to use SMO to train the
                //  underlying SVMs in each of the binary class subproblems.
                Learner = (param) =>new SequentialMinimalOptimization()
                {
                    // Estimate a suitable guess for the Gaussian kernel's
parameters.
                    // This estimate can serve as a starting point for a
grid search.
                    UseKernelEstimation = true
                }
            };
  var calibration = new MulticlassSupportVectorLearning()
            {
                Model = machine, // We will start with an existing machine

                // Configure the learning algorithm to use SMO to train the
                //  underlying SVMs in each of the binary class subproblems.
                Learner = (param) => new ProbabilisticOutputCalibration()
                {
                    Model = param.Model // Start with an existing machine
                }
            };
Please let me know what is the right syntax for this?

Thanks,
Kiran

Best Regards,
Kiran

**Comment by cesarsouza** [comment]
Hi Kiran,

Please, what error are you getting?

Best regards,
Cesar

**Comment by kiran-reddy1** [comment]
Cannot implicitly convert type 'Accord.MachineLearning.VectorMachines.Learning.SequentialMinimalOptimization' to 'Accord.MachineLearning.ISupervisedLearning<Accord.MachineLearning.VectorMachines.SupportVectorMachine<Accord.Statistics.Kernels.IKernel<double[]>>, double[], bool>'.         

'Accord.MachineLearning.VectorMachines.Learning.ProbabilisticOutputCalibration' to 'Accord.MachineLearning.ISupervisedLearning<Accord.MachineLearning.VectorMachines.SupportVectorMachine<Accord.Statistics.Kernels.IKernel<double[]>>, double[], bool>                                                        I tried Gaussian Kernel along with the array inputs. But still getting the errors.

**Comment by cesarsouza** [comment]
Thanks for reporting the issue! Indeed, in order to achieve better performance and type safety the learning algorithms now expect the kernel function to be passed as a generic argument. For example, [please take a look at the third example on the MulticlassSupportVectorLearning documentation](http://accord-framework.net/docs/html/T_Accord_MachineLearning_VectorMachines_Learning_MulticlassSupportVectorLearning.htm):

```c#
// Let's say we have the following data to be classified
// into three possible classes. Those are the samples:
// 
double[][] inputs =
{
    //               input         output
    new double[] { 0, 1, 1, 0 }, //  0 
    new double[] { 0, 1, 0, 0 }, //  0
    new double[] { 0, 0, 1, 0 }, //  0
    new double[] { 0, 1, 1, 0 }, //  0
    new double[] { 0, 1, 0, 0 }, //  0
    new double[] { 1, 0, 0, 1 }, //  1
    new double[] { 0, 0, 0, 1 }, //  1
    new double[] { 0, 0, 0, 1 }, //  1
    new double[] { 1, 0, 1, 1 }, //  2
    new double[] { 1, 1, 0, 1 }, //  2
    new double[] { 0, 1, 1, 1 }, //  2
    new double[] { 1, 1, 1, 1 }, //  2
};

int[] outputs = // those are the class labels
{
    0, 0, 0, 0, 0,
    1, 1, 1, 
    2, 2, 2, 2, 
};

// Create the multi-class learning algorithm for the machine
var teacher = new MulticlassSupportVectorLearning<Gaussian>()
{
    // Configure the learning algorithm to use SMO to train the
    //  underlying SVMs in each of the binary class subproblems.
    Learner = (param) => new SequentialMinimalOptimization<Gaussian>()
    {
        // Estimate a suitable guess for the Gaussian kernel's parameters.
        // This estimate can serve as a starting point for a grid search.
        UseKernelEstimation = true
    }
};

// Learn a machine
var machine = teacher.Learn(inputs, outputs);


// Create the multi-class learning algorithm for the machine
var calibration = new MulticlassSupportVectorLearning<Gaussian>()
{
    Model = machine, // We will start with an existing machine

    // Configure the learning algorithm to use SMO to train the
    //  underlying SVMs in each of the binary class subproblems.
    Learner = (param) => new ProbabilisticOutputCalibration<Gaussian>()
    {
        Model = param.Model // Start with an existing machine
    }
};


// Configure parallel execution options
calibration.ParallelOptions.MaxDegreeOfParallelism = 1;

// Learn a machine
calibration.Learn(inputs, outputs);

// Obtain class predictions for each sample
int[] predicted = machine.Decide(inputs);
```

Please note that the above code is using ```MulticlassSupportVectorLearning<Gaussian>``` instead of just ```MulticlassSupportVectorLearning```. Specifying the kernel as a generic argument might solve the errors you were encountering.

Hope it helps!

Regards,
Cesar

**Comment by Manimojy** [comment]
I gave List<double[]> as input, each array being double[8]. List count is 128. 
And as output I gave 128 labels starting from 0 to 127. 

Following is my code excerpt:

           msvl = new MulticlassSupportVectorLearning<Gaussian>()
            {                
                Learner = (param) => new SequentialMinimalOptimization<Gaussian>()
                {                    
                    UseKernelEstimation = true                    
                }
            };

            msvl.ParallelOptions.MaxDegreeOfParallelism = 1;
            machine0 = msvl.Learn(centroidsAsInput.ToArray(), centroidsLabel);

The program stops at the mvl.Learn with this error:
Error : IndexOutOfRangeException: Index was outside the bounds of the array.

**Comment by Manimojy** [comment]
Solved the IndexOutOfRange exception by giving more input than the number of labels. I had to add a dummy double array for each array of input. :(

Is it a bug?

**Comment by cesarsouza** [comment]
Hi @Manimojy,

Sorry for the delay in answering, sometimes I do not get notifications on existing issues. If it is possible, please share part of the data you are using to train the SVMs. 

Do you have a single example of each class? In this case, the OneVsOne multi-class learning algorithm used by MulticlassSupportVectorLearning will not work. Please try using MultilabelSupportVectorLearning instead.

Regards,
Cesar

**Comment by Manimojy** [comment]
Thanks for the reply. As I said earlier I have solved by adding more data. Since I have updated my app I am not able to reproduce the data; sorry for that. I  will try to create a new issue regarding a new problem in MultiLabelSupportVectorLearning.
btw.. great framework and you have done an AWESOME work. Million thanks to Cesar de Souza. :)

**Comment by cesarsouza** [comment]
Many thanks! I am really glad to know the framework has been useful to you, and I hope it keeps this way :-)

If you had only 128 samples and 128 classes, could it be the case that you had no samples for some of those classes, or just barely (like 1 or 2) for each class? Then in this case it is true that the code might throw errors. I am glad it got solved, in any case!
